docs/src/customization.md

```diff
@@ -74,6 +74,71 @@ There is an example in the [official ST documentation](https://www.sublimetext.c
 
 The following tables give an overview about the scope names used by LSP.
 
+### Semantic Highlighting
+
+!!! info "This feature is only available if the server has the *semanticTokensProvider* capability."
+    Language servers which support semantic highlighting are for example *clangd* and *rust-analyzer*.
+
+In order to support semantic highlighting, the color scheme requires a special rule with a background color set for semantic tokens, which is (marginally) different from the original background.
+LSP automatically adds such a rule to the built-in color schemes from Sublime Text.
+If you use a custom color scheme, select `UI: Customize Color Scheme` from the Command Palette and add for example the following code:
+
+```json
+{
+    "rules": [
+        {
+            "scope": "meta.semantic-token",
+            "background": "#00000101"
+        },
+    ]
+}
+```
+
+Furthermore it is possible to adjust the colors for semantic tokens by applying a foreground color to the individual token types:
+
+| scope | [SemanticTokenTypes](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#semanticTokenTypes) |
+| ----- | ------------------ |
+| `meta.semantic-token.namespace` | namespace |
+| `meta.semantic-token.type` | type |
+| `meta.semantic-token.class` | class |
+| `meta.semantic-token.enum` | enum |
+| `meta.semantic-token.interface` | interface |
+| `meta.semantic-token.struct` | struct |
+| `meta.semantic-token.typeparameter` | typeParameter |
+| `meta.semantic-token.parameter` | parameter |
+| `meta.semantic-token.variable` | variable |
+| `meta.semantic-token.property` | property |
+| `meta.semantic-token.enummember` | enumMember |
+| `meta.semantic-token.event` | event |
+| `meta.semantic-token.function` | function |
+| `meta.semantic-token.method` | method |
+| `meta.semantic-token.macro` | macro |
+| `meta.semantic-token.keyword` | keyword |
+| `meta.semantic-token.modifier` | modifier |
+| `meta.semantic-token.comment` | comment |
+| `meta.semantic-token.string` | string |
+| `meta.semantic-token.number` | number |
+| `meta.semantic-token.regexp` | regexp |
+| `meta.semantic-token.operator` | operator |
+
+By default, LSP will assign scopes based on the [scope naming guideline](https://www.sublimetext.com/docs/scope_naming.html) to each of these token types, but if you define color scheme rules for the scopes specified above, the latter will take precedence.
+
+Language servers can also add their own custom token types, which are not defined in the protocol.
+A "LSP-*" helper package (or user) can provide a "semantic_tokens" mapping in the server configuration for such additional token types.
+Keys of this mapping should be the token types and values should be the corresponding scopes, for example:
+
+```json
+{
+    "semantic_tokens": {
+        "magicFunction": "support.function.builtin",
+        "selfParameter": "variable.language"
+    }
+}
+```
```

Comment:
I've noticed that in VSCode you can map from a token to multiple scopes and also you can set the mapping per language ID.

Example here: https://github.com/johnsoncodehk/volar/blob/b71700d1ce0ff8be7fda857b022530e02a58f2bd/extensions/vscode-vue-language-features/package.json#L318-L351

Would our simpler approach potentially limit us in the future?

Reply:
As mentioned before, the scopes for tokens with modifiers can't be controlled yet. I'm still waiting for a good idea or suggestion about how a proper implementation can work :)
The implementation from the link seems limited to one modifier per token type (?)

I guess the language ID is used in VSCode for the selector in themes, which have it appended after a colon, for example `variable.readonly:java`.
It might be better to use the config name instead, so that tokens from different servers for the same language can be distinguished. But I'm not sure if it is necessary at all, because you could already set it as part of the scope, like `"magicFunction": "support.function.builtin.python"`. It would probably be more important for the default mapping which is defined here in this PR, in case you want different rules for different servers for these tokens in color schemes.

Reply:
I will try to implement something with dotted keys in the token/scope mapping and see if it works. Then we don't need the SEMANTIC_TOKENS_WITH_MODIFIERS_MAP anymore.

Reply:
> The implementation from the link seems limited to one modifier per token type (?)

The values are in a list so I'm assuming that more than one value is supported (even if not used in that example).

> It might be better to use the config name instead, so that tokens from different servers for the same language can be distinguished.

What I was thinking is that having language ID supported might be needed or at least good to have when dealing with a server that supports multiple different languages. Like `Volar` for example which can run both on Vue and JS/TS files). Though I'm not sure sure how likely it would be for this to be an issue.

Reply:
Tokens with at most one modifier are supported now, both in the token/scope map and for color scheme rules. I forgot to add the language ID, but will do it later. I see that the language ID is probably better than the config name.

I don't know why the scopes are in a list in VSCode or how it should work if there are multiple values. Maybe the scopes are just appended in order in that case (but that would already be possible with a single string, with the spaces in scope strings being essential for how selectors work).